Upgrade to polymer 3

cristinecula opened the pull request:
Fixes #12

cristinecula committed:
(dev) remove sync-npm

cristinecula committed:
(wip) run polymer-modulizer

1. install pristine `bower_components` and `node_modules`

`rm -rf node_modules bower_components && yarn install`

2. install modulizer

`npm i -g polymer-modulizer`

Gotcha 1: Modulizer only detects the template if html and js files use 2 space indentation.

3. convert tabs to spaces using find and expand:

`find . \( -name '*.html' -o -name "*.js" \) -not -path "./node_modules/*" -not -path "./bower_components/*" -not -path "./coverage/*" -exec bash -c 'expand -t 2 "$0" > /tmp/e && mv /tmp/e "$0"' {} \;`

4. run modulizer. Use force because repo is dirty.

`modulizer --out . --force`

5. convert back spaces to tabs using unexpand:

`find . \( -name '*.html' -o -name "*.js" \) -not -path "./node_modules/*" -not -path "./bower_components/*" -not -path "./coverage/*" -exec bash -c 'unexpand -t 2 "$0" > /tmp/e && mv /tmp/e "$0"' {} \;`

cristinecula committed:
(wip) install new packages

1. remove polymer install hook from `package.json`

`"postinstall": "polymer install",`

2. remove `bower.json`

`rm -rf bower_components bower.json`

3. remove files and folders relating to `bower` from `.gitignore`

4. install the new packages and commit the lockfile

`yarn`

cristinecula committed:
(wip) make `polymer serve` resolve components from `node_modules`

cristinecula committed:
(wip) update linter config

cristinecula committed:
(wip) add missing dependencies

Sometimes `polymer-modulizer` cannot guess how to map a bower lib to a npm package.
If you run `yarn lint` it should list missing dependencies.

cristinecula committed:
(wip) `yarn add` re-orders the packages alphanumerically

cristinecula committed:
(wip) update `yarn analyze` command and fix doc comments

cristinecula committed:
(wip) neon-animation does not include the web animations polyfill anymore.

cristinecula committed:
(wip) cleanup README.md

cristinecula committed:
(wip) demo does not work

cristinecula committed:
(wip) clean up demo

cristinecula committed:
(wip) upgrade test runner config

`yarn remove wct-browser-legacy web-component-tester-istanbul`
`yarn add wct-mocha mocha@^5.0.0 chai @polymer/test-fixture wct-istanbul --dev`

cristinecula committed:
(wip) fix tests

cristinecula committed:
(wip) use absolute path for imports, remove duplicated imports

cristinecula committed:
(wip) update travis config

cristinecula committed:
(wip) do not test on browsers that fail

The tests fail on Firefox and Edge.

Reason:
`polymer serve` tries to support all browsers by transforming javascript modules to AMD modules. The bad news is that `moment` has a circular dependency which causes the module to not initialize properly (AMD does not support circular dependencies). Firefox and Edge trigger the modules transformation, because they do not support javascript modules.

The good news is that Firefox added support in v67 and Edge is switching to chromium anyway.

When https://github.com/Polymer/tools/pull/3420 is released, we should add back Firefox to the tested browsers.

cristinecula commented:
@megheaiulian @nomego this should be good to go :)

cristinecula commented:
@nomego To be able to include this component in the other components, it needs to be published to npm. What package name do you want to use? Do you want to leave it as `cosmoz-moment` or "move" it under the organization `@neovici/cosmoz-moment`?

cristinecula committed:
(wip) upgrade analysis.json

`yarn analyze`

cristinecula commented:
I updated the package name to `@neovici/cosmoz-moment` and set the version to `3.0.0`.

cristinecula committed:
(wip) paths should be absolute

cristinecula committed:
(wip) sort package.json

nomego commented on cosmoz-moment-timeago.js:
```diff
@@ -0,0 +1,72 @@
+	Cosmoz.MomentBehavior
+], PolymerElement) {
+	static get template() {
+		return html`
```
Why not a a single line ?

nomego commented on cosmoz-moment-timeago.js:
```diff
@@ -0,0 +1,72 @@
+
+	static get observers() {
+		return [
+			'refreshIntervalChanged(refreshInterval)'
```
Should be defined as an `observer` function within the property declaration.

nomego commented on cosmoz-moment-timeago.js:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	refreshIntervalChanged(refreshInterval) {
+		this.set('_intervalId', window.setInterval(function () {
```
Arrow function ?

nomego commented on cosmoz-moment.js:
```diff
@@ -0,0 +1,109 @@
+
+	static get observers() {
+		return [
+			'localeChanged(locale)'
```
Should also be an observer function in the property declaration.

nomego reviewed (approved):
Can be merged as is but there are some stuff that should be captured for future changes.
(The PR really is about porting and I assume everything commented on is like this right now)
Also we should add a `Cosmoz.Mixins.moment()` mixin to use instead of using the behavior directly.

cristinecula committed:
(wip) upgrade eslint and it's plugins

`yarn upgrade eslint-plugin-mocha@^5.0.0 eslint@^5.0.0 eslint-plugin-html@^5.0.0`

cristinecula committed:
(wip) use @neovici eslint config

cristinecula committed:
(wip) fix lint errors

cristinecula committed:
(wip) add firefox back to wct config

cristinecula committed:
(wip) remove resolutions; update eslint-config

cristinecula committed:
(wip) refactor MomentBehavior to momentAware mixin

cristinecula committed:
(wip) update analysis

cristinecula committed:
(wip) travis: xenial builds faster

cristinecula committed:
(wip) should also be an observer function in the property declaration.

cristinecula committed:
(wip) Should be defined as an observer function within the property declaration.

cristinecula commented:
@nomego it appears that `polymer-analyzer` (which is used by `polymer lint`) expects that the observer descriptor is always a string. This makes the linter fail with observers defined in the way documented in https://polymer-library.polymer-project.org/2.0/docs/devguide/observers#simple-observers:

```
  static get properties() {
    return {
      active: {
        type: Boolean,
        // Observer method
        observer(newValue, oldValue){
          this.toggleClass('highlight', newValue);
        }
      }
    }
  }
```

If you go to the [3.0 documentation](https://polymer-library.polymer-project.org/3.0/docs/devguide/observers#simple-observers) you can see that the section has been removed.

cristinecula committed:
(wip) import paths can omit .js

cristinecula committed:
Revert "(wip) should also be an observer function in the property declaration."

This reverts commit 3256a2bcae4a7f5ddfe2c0130e41d0acb68993c6.

cristinecula committed:
Revert "(wip) Should be defined as an observer function within the property declaration."

This reverts commit d1799e8cd3459814f9daeab7715a3150a6388660.

cristinecula commented:
@nomego ready to release!
